x/upgrade/keeper/keeper.go

```diff
@@ -177,107 +196,135 @@ func (k Keeper) GetModuleVersions(ctx sdk.Context) []*types.ModuleVersion {
-			return version, true
+			return version, true, nil
 		}
 	}
-	return 0, false
+
+	return 0, false, it.Close()
 }
 
 // ScheduleUpgrade schedules an upgrade based on the specified plan.
 // If there is another Plan already scheduled, it will cancel and overwrite it.
 // ScheduleUpgrade will also write the upgraded IBC ClientState to the upgraded client
 // path if it is specified in the plan.
-func (k Keeper) ScheduleUpgrade(ctx sdk.Context, plan types.Plan) error {
+func (k Keeper) ScheduleUpgrade(ctx context.Context, plan types.Plan) error {
 	if err := plan.ValidateBasic(); err != nil {
 		return err
 	}
 
 	// NOTE: allow for the possibility of chains to schedule upgrades in begin block of the same block
 	// as a strategy for emergency hard fork recoveries
-	if plan.Height < ctx.BlockHeight() {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	if plan.Height < sdkCtx.BlockHeight() {
 		return errorsmod.Wrap(sdkerrors.ErrInvalidRequest, "upgrade cannot be scheduled in the past")
 	}
 
-	if k.GetDoneHeight(ctx, plan.Name) != 0 {
+	doneHeight, err := k.GetDoneHeight(ctx, plan.Name)
+	if err != nil {
+		return err
+	}
+
+	if doneHeight != 0 {
 		return errorsmod.Wrapf(sdkerrors.ErrInvalidRequest, "upgrade with name %s has already been completed", plan.Name)
 	}
 
-	store := ctx.KVStore(k.storeKey)
+	store := k.storeService.OpenKVStore(ctx)
 
 	// clear any old IBC state stored by previous plan
-	oldPlan, found := k.GetUpgradePlan(ctx)
+	oldPlan, found, err := k.GetUpgradePlan(ctx)
+	if err != nil {
+		return err
+	}
+
 	if found {
-		k.ClearIBCState(ctx, oldPlan.Height)
+		err = k.ClearIBCState(ctx, oldPlan.Height)
+		if err != nil {
+			return err
+		}
 	}
 
-	bz := k.cdc.MustMarshal(&plan)
-	store.Set(types.PlanKey(), bz)
+	bz, err := k.cdc.Marshal(&plan)
+	if err != nil {
+		return err
+	}
+
+	err = store.Set(types.PlanKey(), bz)
+	if err != nil {
+		return err
+	}
 
 	telemetry.SetGaugeWithLabels([]string{"server", "info"}, 1, []metrics.Label{telemetry.NewLabel("upgrade_height", strconv.FormatInt(plan.Height, 10))})
 
 	return nil
 }
 
 // SetUpgradedClient sets the expected upgraded client for the next version of this chain at the last height the current chain will commit.
-func (k Keeper) SetUpgradedClient(ctx sdk.Context, planHeight int64, bz []byte) error {
-	store := ctx.KVStore(k.storeKey)
-	store.Set(types.UpgradedClientKey(planHeight), bz)
-	return nil
+func (k Keeper) SetUpgradedClient(ctx context.Context, planHeight int64, bz []byte) error {
+	store := k.storeService.OpenKVStore(ctx)
+	return store.Set(types.UpgradedClientKey(planHeight), bz)
 }
 
 // GetUpgradedClient gets the expected upgraded client for the next version of this chain
-func (k Keeper) GetUpgradedClient(ctx sdk.Context, height int64) ([]byte, bool) {
-	store := ctx.KVStore(k.storeKey)
-	bz := store.Get(types.UpgradedClientKey(height))
-	if len(bz) == 0 {
-		return nil, false
+func (k Keeper) GetUpgradedClient(ctx context.Context, height int64) ([]byte, bool, error) {
```

Comment:
Same, shouldn't it only return an error instead of a bool?

Reply:
I'll return `errors.ErrNotFound` 👌